From 424bed783bc8db0ce7201e0147a01aa26809ba43 Mon Sep 17 00:00:00 2001
From: Yannick Fertre <yannick.fertre@foss.st.com>
Date: Fri, 29 Mar 2024 14:04:41 +0100
Subject: [PATCH 1058/1141] drm/stm : lvds: get pixel clock rate

At probe, get the pixel clock rate from registers settings.

Change-Id: I9acd397e1ee33ef81e3962b0daf120fc9a388130
Signed-off-by: Yannick Fertre <yannick.fertre@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/370226
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/gpu/drm/stm/lvds.c | 26 +++++++++++++++++++++++++-
 1 file changed, 25 insertions(+), 1 deletion(-)

--- a/drivers/gpu/drm/stm/lvds.c
+++ b/drivers/gpu/drm/stm/lvds.c
@@ -1090,6 +1090,9 @@ static int lvds_probe(struct platform_de
 	struct device_node *port1, *port2, *remote;
 	struct stm_lvds *lvds;
 	struct reset_control *rstc;
+	struct lvds_phy_info *phy;
+	unsigned int pll_in_khz, bdiv, mdiv, ndiv;
+	int multiplier, rate;
 	int ret, dual_link;
 
 	DRM_DEBUG_DRIVER("Probing LVDS driver...\n");
@@ -1229,9 +1232,30 @@ static int lvds_probe(struct platform_de
 	 * To obtain a continuous display after the probe,
 	 *  the clocks must remain activated
 	 */
-	if (device_property_read_bool(dev, "default-on"))
+	if (device_property_read_bool(dev, "default-on")) {
 		pm_runtime_get_sync(dev);
 
+		if (lvds->phy_master) {
+			if (lvds_is_dual_link(lvds->config.link_type))
+				multiplier = 2;
+			else
+				multiplier = 1;
+
+			phy = lvds->phy_master;
+			pll_in_khz = clk_get_rate(lvds->pllref_clk) / 1000;
+
+			ndiv = lvds_read(lvds, phy->base + phy->ofs.PxPLLCR2) >> 16;
+			bdiv = lvds_read(lvds, phy->base + phy->ofs.PxPLLCR2) & 0xFFFF;
+			mdiv = lvds_read(lvds, phy->base + phy->ofs.PxPLLSDCR1);
+
+			/* X7 because for each pixel in 1 lane there is 7 bits
+			 * We want pixclk, not bitclk
+			 */
+			rate = pll_get_clkout_khz(pll_in_khz, bdiv, mdiv, ndiv);
+			lvds->pixel_clock_rate = (unsigned long) rate  * 1000 * multiplier / 7;
+		}
+	}
+
 	return 0;
 }
 
